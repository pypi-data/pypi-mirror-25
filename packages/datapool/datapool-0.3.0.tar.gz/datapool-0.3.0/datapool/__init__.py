# -*- coding: utf-8 -*-

__author__ = 'Uwe Schmitt'
__email__ = 'uwe.schmitt@id.ethz.ch'
__version__ = '0.3.0'
