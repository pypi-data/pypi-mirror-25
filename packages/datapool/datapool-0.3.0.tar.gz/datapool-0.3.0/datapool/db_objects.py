# encoding: utf-8
from __future__ import print_function, division, absolute_import

from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import (Column, Integer, String, LargeBinary, ForeignKey, Float, DateTime,
                        UniqueConstraint, Date)
from sqlalchemy.orm import relationship, backref
from datetime import datetime

Base = declarative_base()


class SiteDbo(Base):

    __tablename__ = 'site'

    site_id = Column(Integer(), primary_key=True)
    name = Column(String(), unique=True, index=True, nullable=False)
    description = Column(String())
    street = Column(String())
    postcode = Column(String())
    city = Column(String())
    coord_x = Column(String())
    coord_y = Column(String())
    coord_z = Column(String())

    def __str__(self):
        return ("<SiteDbo name={name}, city={city}, x={coord_x}, y={coord_y}, "
                "z={coord_z}, description={description}>").format(**vars(self))


class SignalDbo(Base):
    """(formerly known as fact table)
    Contains the measured value of a given parameter taken a at a specific time, site
    """

    __tablename__ = 'signal'

    signal_id = Column(Integer(), primary_key=True)
    value = Column(Float(), nullable=False)
    timestamp = Column(DateTime(), nullable=False)

    parameter_id = Column(ForeignKey("parameter.parameter_id"), nullable=False)
    source_id = Column(ForeignKey("source.source_id"), nullable=False)
    site_id = Column(ForeignKey("site.site_id"), nullable=True)

    coord_x = Column(String())
    coord_y = Column(String())
    coord_z = Column(String())

    site = relationship("SiteDbo", backref=backref("signals"))
    source = relationship("SourceDbo", backref=backref("signals"))
    parameter = relationship("ParameterDbo", backref=backref("signals"))

    def __str__(self):
        return ("<SourceDbo signal_id={signal_id}, parameter_id={parameter_id}, "
                "value={value}, timestamp={timestamp}, site={site_id}>").format(**vars(self))


class PictureDbo(Base):

    __tablename__ = 'picture'

    picture_id = Column(Integer(), primary_key=True)
    site_id = Column(ForeignKey("site.site_id"), nullable=False)
    filename = Column(String(), nullable=False)
    description = Column(String())
    timestamp = Column(DateTime())
    data = Column(LargeBinary())

    site = relationship("SiteDbo", backref=backref("pictures"))
    constraint = UniqueConstraint('site_id', 'filename')

    def __str__(self):
        return ("<PictureDbo filename={filename}, timestamp={timestamp}, "
                "description={description}>").format(**vars(self))


class SourceDbo(Base):

    __tablename__ = 'source'

    source_id = Column(Integer(), primary_key=True)
    source_type_id = Column(ForeignKey("source_type.source_type_id"), nullable=False)
    name = Column(String(), unique=True, index=True, nullable=False)
    description = Column(String())
    serial = Column(String())
    manufacturer = Column(String())
    manufacturing_date = Column(Date())

    source_type = relationship("SourceTypeDbo", backref=backref("sources"))

    def __str__(self):
        return ("<SourceDbo name={name}, serial={serial}, "
                "description={description}>").format(**vars(self))


class SourceTypeDbo(Base):

    __tablename__ = 'source_type'

    source_type_id = Column(Integer(), primary_key=True)
    name = Column(String(), unique=True, index=True, nullable=False)
    description = Column(String())

    def __str__(self):
        return "<SourceTypeDbo name={name}, description={description}>".format(**vars(self))


class SpecialValueDefinitionDbo(Base):

    __tablename__ = 'special_value_definition'

    special_value_definition_id = Column(Integer(), primary_key=True)
    source_type_id = Column(ForeignKey("source_type.source_type_id"), nullable=False)
    description = Column(String())
    categorical_value = Column(String(), nullable=False)
    numerical_value = Column(Float(), nullable=False)

    source_type = relationship("SourceTypeDbo", backref=backref("special_values"))

    def __str__(self):
        return ("<SpecialValueDefinitionDbo categorical_value={categorical_value}, "
                "numerical_value={numerical_value}>").format(**vars(self))


class ParameterDbo(Base):

    __tablename__ = 'parameter'

    parameter_id = Column(Integer(), primary_key=True)
    name = Column(String(), unique=True, index=True, nullable=False)
    description = Column(String())
    unit = Column(String(), nullable=False)

    def __str__(self):
        return ("<ParameterDbo name={name}, unit={unit}, description={description}>"
                .format(**vars(self)))


class ParameterAveragingDbo(Base):

    __tablename__ = 'parameter_averaging'

    parameter_averaging_id = Column(Integer(), primary_key=True)
    parameter_id = Column(ForeignKey("parameter.parameter_id"), nullable=False)
    source_id = Column(ForeignKey("source.source_id"), nullable=False)
    integration_length_x = Column(Float())
    integration_length_y = Column(Float())
    integration_angle = Column(Float())
    integration_time = Column(Float())

    constraint = UniqueConstraint('parameter_id', 'source_id')

    parameter = relationship("ParameterDbo", backref=backref("averaging"))
    source = relationship("SourceDbo", backref=backref("averaging"))


class CommentDbo(Base):

    __tablename__ = 'comment'

    comment_id = Column(Integer(), primary_key=True)
    signal_id = Column(ForeignKey("signal.signal_id"), nullable=False)
    text = Column(String(), nullable=False)
    timestamp = Column(DateTime(), nullable=False, default=datetime.now())
    author = Column(String())

    signals = relationship("SignalDbo", backref=backref("comments"))


class SignalQualityDbo(Base):

    __tablename__ = 'signal_quality'

    signal_quality_id = Column(Integer(), primary_key=True)
    signal_id = Column(ForeignKey("signal.signal_id"), nullable=False)
    quality_id = Column(ForeignKey("quality.quality_id"), nullable=False)
    timestamp = Column(DateTime(), nullable=False, default=datetime.now())
    author = Column(String())


class QualityDbo(Base):

    __tablename__ = 'quality'

    quality_id = Column(Integer(), primary_key=True)
    flag = Column(String(), nullable=False)
    method = Column(String())


class ProjectDbo(Base):

    __tablename__ = 'project'

    project_id = Column(Integer(), primary_key=True)
    title = Column(String(), nullable=False)
    description = Column(String())
