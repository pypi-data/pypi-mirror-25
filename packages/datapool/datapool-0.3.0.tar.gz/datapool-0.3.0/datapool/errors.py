# encoding: utf-8
from __future__ import print_function, division, absolute_import


class DataPoolException(Exception):
    pass


class ConsistencyError(DataPoolException):
    pass


class FormatError(DataPoolException):
    pass


class InvalidOperationError(DataPoolException):
    pass


class InvalidScriptName(DataPoolException):
    pass


class InvalidLandingZone(DataPoolException):
    pass


class InvalidConfiguration(DataPoolException):
    pass


class DataPoolIOError(DataPoolException):
    pass
