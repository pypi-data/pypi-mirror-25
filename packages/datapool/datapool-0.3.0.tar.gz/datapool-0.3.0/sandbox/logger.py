# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datetime import datetime as dt
from functools import partial

import zmq


class Logger:
    """This is sending signals to the log_receiver via port 5559 (using zmq)
    """

    def __init__(self, context, id_=""):
        self.id_ = id_
        self.sender = context.socket(zmq.PUSH)
        self.sender.connect("tcp://127.0.0.1:5559")

        self.debug = partial(self._log, "debug")
        self.info = partial(self._log, "info")
        self.warn = partial(self._log, "warn")
        self.error = partial(self._log, "error")

    def shutdown(self):
        self.sender.send(b"KILL|")

    def _log(self, level, *a):
        text = " ".join(map(str, a))
        #text = "%-12s: %s" % (self.prefix, text)
        message = "%s|%s|%s|%s" % (level, dt.now(), self.id_, text)
        self.sender.send(message.encode("utf-8"), flags=zmq.NOBLOCK)
