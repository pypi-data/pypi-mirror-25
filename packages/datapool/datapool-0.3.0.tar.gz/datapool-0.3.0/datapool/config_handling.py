# encoding: utf-8
from __future__ import print_function, division, absolute_import

from collections import OrderedDict
from contextlib import closing
import os
import shutil

from .config import MagicConfig, write_ini, read_ini
from .errors import InvalidOperationError
from .utils import find_executable, iter_to_list


def guess_config(known_config):
    """you can provide known config settings as

       known_config = { "db.connection_string:"...",
                        "worker.port": 3333 }

       which will overrung the settings we guess in this function
    """

    assert isinstance(known_config, dict)

    executables = ("matlab", "R", "julia", "python3")
    sections = ("matlab", "r", "julia", "python")

    messages = []

    # we use OrderedDict to guarantee stable ordering of entries in config
    # file(s)
    config = OrderedDict()
    for executable, section in zip(executables, sections):
        path = find_executable(executable, "")
        if not path:
            messages.append("{!r} not found on $PATH".format(executable))
        config["{}.executable".format(section)] = path

    config.update(known_config)

    return default_config(config), messages


def default_config(known_config):
    """you can overwrite with calls like:

        default_config(db.connections_string="....",
                       worker.port=3333)
    """

    config = MagicConfig()

    config.db.connection_string = "postgresql://user:password@127.0.0.1:5432/datapool"

    config.backup_landing_zone.folder = ""

    config.r.extension = ".r"

    config.matlab.extension = ".m"

    config.julia.extension = ".jl"
    config.julia.version = "0.5.0"

    config.python.extension = ".py"

    config.logging.config_file = "./logging_config.yaml"
    config.log_receiver.port = 5559

    config.worker.port = 5555
    config.worker.count = 5

    for key, value in known_config.items():
        fields = key.split(".")
        assert len(fields) == 2
        section, field = fields
        setattr(getattr(config, section), field, value)

    return config


def config_for_develop_db(landing_zone_folder):
    config = MagicConfig()
    path = os.path.join(landing_zone_folder, ".develop.db")
    config.connection_string = "sqlite+pysqlite:///{}".format(path)
    return config, path


def check_folder(folder):

    if not os.path.exists(folder):
        return "folder '{}' does not exist".format(folder)

    try:
        existing_files = os.listdir(folder)
    except IOError:
        return "can list content of folder '{}'".format(folder)

    existing_files = [f for f in existing_files if not f.startswith(".")]

    if existing_files:
        return "folder is not empty '{}'".format(folder)

    marker_file = os.path.join(folder, ".marker")
    try:
        open(marker_file, "w").close()
    except IOError:
        return "can not write to folder '{}'".format(folder)

    os.remove(marker_file)


@iter_to_list
def check_config(config):

    from .database import connect_to_db

    @iter_to_list
    def check(section, field, type, maybe_empty=True):
        if section not in config.keys():
            yield False, "- section {} missing".format(section)
        if field not in config[section].keys():
            yield False, "- field {} in section {} missing".format(field, section)

        value = config[section][field]
        if not maybe_empty and value == "":
            yield False, "- field {} in section {} has no value set".format(field, section)
        if not isinstance(value, type):
            yield (False, "- field {} in section {} is not of type {}".format(field, section,
                                                                              type.__name__))

    errors = check("landing_zone", "folder", str, False)
    yield from errors

    if not errors and config.landing_zone.folder:
        yield True, "- check landing zone '{}'".format(config.landing_zone.folder)
        msg = check_folder(config.landing_zone.folder)
        if msg is not None:
            yield False, "- {}".format(msg)

    errors = check("backup_landing_zone", "folder", str, True)
    yield from errors

    if not errors and config.backup_landing_zone.folder:
        yield True, "- check backup folder '{}'".format(config.backup_landing_zone.folder)
        msg = check_folder(config.backup_landing_zone.folder)
        if msg is not None:
            yield False, "- {}".format(msg)

    errors = check("db", "connection_string", str, False)
    yield from errors

    if not errors:
        try:
            yield True, "- try to connect to db"
            connect_to_db(config.db)
            yield True, "  - connected to db"
        except Exception as e:
            yield False, "  - {}".format(e)

    errors = check("r", "extension", str, False)
    yield from errors

    errors = check("r", "executable", str)
    yield from errors

    if not errors:
        if config.r.executable == "":
            yield True, "- R not configured, skip tests"
        else:
            yield True, "- check R configuration + code execution"
            from .r_runner import RRunner
            m = RRunner(config.r.executable)
            try:
                m.start_interpreter()
                yield True, "- R interpreter works"
            except:
                yield False, "  - could not start R from {}".format(config.r.executable)

    errors = check("matlab", "extension", str, False)
    yield from errors

    errors = check("matlab", "executable", str)
    yield from errors
    if not errors:
        if config.matlab.executable == "":
            yield True, "- matlab not configured, skip tests"
        else:
            yield True, "- check matlab configuration + code execution"
            from .matlab_runner import MatlabRunner
            m = MatlabRunner(config.matlab.executable)
            try:
                m.start_interpreter()
                yield True, "- matlab interpreter works"
            except:
                yield False, "  - could not start matlab from {}".format(config.matlab.executable)

    errors = check("julia", "extension", str, False)
    yield from errors

    errors = check("julia", "executable", str)
    yield from errors

    if not errors:
        if config.julia.executable == "":
            yield True, "- julia not configured, skip tests"
        else:
            yield True, "- check julia configuration + code execution"
            from .julia_runner import JuliaRunner
            m = JuliaRunner(config.julia.executable)
            try:
                m.start_interpreter()
                yield True, "- julia interpreter works"
            except:
                yield False, "  - could not start julia from {}".format(config.julia.executable)
            else:
                version = m.get_julia_version_string()
                if config.julia.version != version:
                    yield False, ("- julia interpreter is of version {}, configured is {}"
                                  "".format(version, config.julia.version))
                else:
                    yield True, "- checked julia version."

    errors = check("python", "extension", str, False)
    yield from errors

    errors = check("python", "executable", str)
    yield from errors

    if not errors:
        if config.python.executable == "":
            yield True, "- python not configured, skip tests"
        else:
            yield True, "- check python configuration + code execution"
            from .python_runner import PythonRunner
            m = PythonRunner(config.python.executable)
            try:
                m.start_interpreter()
                yield True, "- python interpreter works"
            except:
                yield False, "  - could not start python from {}".format(config.python.executable)

    errors = check("logging", "config_file", str, "")
    yield from errors

    # local import avoids circular import here:
    from .logger import resolve_logger_config_path
    if not errors and not os.path.exists(resolve_logger_config_path(config)):
        yield (False, "- configured config logging config {} does not exist".format(
            config.logging.config_file))
    yield from errors

    errors = check("log_receiver", "port", int)
    yield from errors

    if not errors:
        port = config.log_receiver.port
        if _is_port_used(port):
            yield False, "- configured log_receiver port {} is already in use".format(port)

    errors = check("worker", "port", int)
    yield from errors

    if not errors:
        port = config.worker.port
        if _is_port_used(port):
            yield False, "- configured worker port {} is already in use".format(port)

    errors = check("worker", "count", int)
    yield from errors
    if not errors and config.worker.count <= 0:
        yield (False, "- configured worker count value {} is not greater than zero".format(
            config.worker.count))


def _is_port_used(port):
    import socket
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        sock.settimeout(.5)
        return sock.connect_ex(('127.0.0.1', port)) == 0


def datapool_folder():
    etc = os.environ.get("ETC", "/etc")
    data_pool_folder = os.path.join(etc, "datapool")
    return os.path.abspath(data_pool_folder)


def datapool_ini_file_path():
    folder = datapool_folder()
    path = os.path.join(folder, "datapool.ini")
    return path


def init_config(landing_zone_folder, sqlite_db=False, overwrite=False):

    config_folder = datapool_folder()

    if os.path.exists(config_folder):
        if not overwrite:
            raise InvalidOperationError("datapool folder {} already exists".format(config_folder))
    else:
        try:
            os.makedirs(config_folder)
        except PermissionError:
            raise InvalidOperationError("creation of {} failed. try sudo.".format(config_folder))

    path = datapool_ini_file_path()

    # we use OrderedDict to guarantee stable ordering of entries in config
    # file(s)
    known_settings = OrderedDict([("landing_zone.folder", landing_zone_folder)])
    if sqlite_db:
        db_path = os.path.join(landing_zone_folder, ".simple.db")
        known_settings["db.connection_string"] = "sqlite+pysqlite:///{}".format(db_path)

    config, messages = guess_config(known_settings)

    try:
        write_ini(config, path)
    except PermissionError:
        raise InvalidOperationError("could not write {}. try sudo.".format(path))

    from .utils import abs_folder
    default_log_config = os.path.join(abs_folder(__file__), "cmdline_logging.yaml")
    shutil.copy(default_log_config, os.path.join(config_folder, config.logging.config_file))

    return config_folder, messages


def read_config(**variable_settings):
    path = datapool_ini_file_path()
    if os.path.exists(path):
        return read_ini(path, variable_settings)
    else:
        return None


def write_config(config):
    assert "__file__" in config.keys()
    write_ini(config, config.__file__)
