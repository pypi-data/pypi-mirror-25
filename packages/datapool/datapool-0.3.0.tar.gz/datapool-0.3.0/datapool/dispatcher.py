# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datetime import datetime
import functools
import os
import shutil
import time

from .database import connect_to_db
from .data_conversion import ConversionRunner, SUPPORTED_EXTENSIONS
from .errors import DataPoolException, InvalidOperationError
from .domain_objects import DomainObject
from .landing_zone_structure import (lookup_handler, script_paths_for_raw_file,
                                     source_yaml_path_for_script)

from .logger import logger
from .signal_model import commit as commit_signals, check_signals_against_db
from .uniform_file_format import to_signals
from .utils import iter_to_list
from .yaml_parsers import parse_source


def log_unexpected_exceptions(function):
    @functools.wraps(function)
    def wrapped_function(*a, **kw):
        logger()
        try:
            return function(*a, **kw)
        except Exception as e:
            args = list("{!r}".format(arg) for arg in a)
            args += list("{}={!r}".format(k, v) for (k, v) in sorted(kw.items()))
            call_signature = "{}({})".format(function.__name__, ", ".join(args))
            logger().error("unexpected error '{}' when calling {}".format(e, call_signature))

    return wrapped_function


def default_time_provider():
    return datetime.now().strftime("%Y%m%d-%H%M%S-%f")


class Dispatcher:

    def __init__(self, config, time_provider=default_time_provider):
        """
        time_provider arg allows patching for regression tests !
        """
        self.config = config
        self.time_provider = time_provider
        self.root_folder = config.landing_zone.folder
        self.conversion_runner = ConversionRunner(config)
        self.engine = connect_to_db(self.config.db)

    @iter_to_list
    def dispatch(self, rel_path):
        if rel_path.endswith(".yaml"):
            yield from self.dispatch_yaml(rel_path)
        elif rel_path.endswith(".raw"):
            yield from self.dispatch_raw_file_conversion(rel_path)
        elif any(rel_path.endswith(ext) for ext in SUPPORTED_EXTENSIONS):
            # no dispatch because dispatch is triggered on .raw file creation, only backup:
            yield from self.backup(rel_path)
        else:
            yield "do not know how to process file {}".format(rel_path)

    def backup(self, rel_path):
        try:
            yield from self._backup(rel_path)
        except IOError as e:
            yield "backup of {} failed: {}".format(rel_path, e)

    def _backup(self, rel_path):
        backup_path = os.path.join(self.config.backup_landing_zone.folder, rel_path)
        folder = os.path.dirname(backup_path)
        if not os.path.exists(folder):
            os.makedirs(folder)
            yield "created folder {}".format(folder)

        # add time stamp to make file unique and better searchable:
        backup_path += ".{}".format(self.time_provider())

        shutil.copy(self.full_path(rel_path), backup_path)
        yield "copied {} to {}".format(rel_path, backup_path)

    def full_path(self, *a):
        return os.path.join(self.root_folder, *a)

    def rel_path(self, p):
        return os.path.relpath(p, self.root_folder)

    def dispatch_raw_file_conversion(self, rel_raw_file_path):
        raw_file_path = self.full_path(rel_raw_file_path)
        script_paths = script_paths_for_raw_file(raw_file_path)

        if len(script_paths) > 1:
            yield InvalidOperationError("multiple conversion scripts found for {}"
                                        .format(rel_raw_file_path))
            return

        if len(script_paths) == 0:
            yield InvalidOperationError("no conversion script found next to {}"
                                        .format(rel_raw_file_path))
            return

        yield from self.run_conversion(script_paths[0], raw_file_path)
        yield from self.backup(rel_raw_file_path)

    def run_conversion(self, script_path, raw_file_path):
        rel_script_path = self.rel_path(script_path)
        rel_raw_file_path = self.rel_path(raw_file_path)

        msg = "convert {} with {}".format(rel_raw_file_path, rel_script_path)
        yield msg
        logger().info(msg)

        started = time.time()
        try:
            rows = self.conversion_runner.run_conversion(script_path, raw_file_path)
        except DataPoolException as e:
            msg = "converting {} with {} failed: '{}'".format(rel_raw_file_path, rel_script_path, e)
            yield msg
            logger().error(msg)
            return
        needed = time.time() - started

        msg = "timing: converting {} needed {:.2f} seconds".format(rel_raw_file_path, needed)
        yield msg
        logger().info(msg)

        source = next(parse_source(self.root_folder, source_yaml_path_for_script(script_path)))
        for row in rows:
            row["source"] = source.name

        started = time.time()
        signals = to_signals(rows)
        errors = check_signals_against_db(signals, self.engine)
        if errors:
            MAXE = 50
            too_many = len(errors) > MAXE
            orig_len = len(errors)
            errors = errors[:MAXE]
            for e in errors:
                msg = "converting {} with {} failed: '{}'".format(rel_raw_file_path,
                                                                  rel_script_path, e)
                yield InvalidOperationError(msg)
                logger().error(msg)
            if too_many:
                msg = "skipped {} error messages".format(orig_len - MAXE)
                yield InvalidOperationError(msg)

            return
        msg = "checked {} signals, needed {:.2f} seconds".format(len(signals), needed)
        yield msg
        logger().info(msg)

        started = time.time()
        commit_signals(signals, self.engine)
        needed = time.time() - started
        msg = "commited {} signals, needed {:.2f} seconds".format(len(signals), needed)
        logger().info(msg)
        yield msg

    def dispatch_yaml(self, rel_path):
        handler = lookup_handler(rel_path)

        if handler is None:
            yield InvalidOperationError("don't know how to handle {}".format(rel_path))
            return

        ok = True
        domain_object = None
        for result in handler.parser(self.root_folder, rel_path):
            if isinstance(result, DomainObject):
                domain_object = result
                continue
            yield InvalidOperationError("when parsing {}: {}".format(rel_path, result))
            ok = False

        if not ok:
            return

        if domain_object is None:
            yield "nothing to commit"
            return

        try:
            committed = handler.committer(domain_object, self.engine)
            if committed:
                yield "commited changes"
            else:
                yield "nothing committed, possibly no update"
        except DataPoolException as e:
            yield e

        self.backup(rel_path)
