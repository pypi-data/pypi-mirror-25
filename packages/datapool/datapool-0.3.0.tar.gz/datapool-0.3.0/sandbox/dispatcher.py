# encoding: utf-8

import os
import signal
import subprocess
import sys
import threading
import time

import zmq

from logger import Logger


# for debugging: print stacktrace if SIGUSR1 is received:
import faulthandler
faulthandler.register(signal.SIGUSR1)


def run_worker(*args):
    # it is better to use absolution imports here, but for development this will not work
    # unless we implement a command line to start the dispatcher
    try:
        from . import worker
    except:
        import worker
    worker_path = os.path.abspath(worker.__file__)
    cmdline = [sys.executable, worker_path] + list(args)

    def preexec():
        signal.signal(signal.SIGINT, signal.SIG_IGN)

    # here the argument preexec_fn avoids forwarding of KeyboardInterrupt to workers
    return subprocess.Popen(cmdline, bufsize=0, preexec_fn=preexec)


class Dispatcher:
    """ The dispatcher is receiving work from demo.py (aka the client) via 'inproc://client' (in-process communication)
    A poller is introduced, which watches both for incoming work and available workers (communicating via port 5555).
    If there is incoming work, a worker is popped from the list of available workers and the work is sent to that worker.
    When the worker is finished, it sends the message "READY" back to the poller; the worker is then added to the list of available workers again.

    
    At any time, log-info is sent to the logger which is sending the message to the log_receiver (port 5559)
    """

    def __init__(self, num_workers, context, worker_ip="127.0.0.1", worker_port=5555):
        self.num_workers = num_workers

        self.workers_connect_to = "tcp://%s:%d" % (worker_ip, worker_port)
        self.client_sends_to = "inproc://client"

        self.context = context
        self.logger = Logger(context, "dispatcher")

        self._start_workers()
        self._setup_client_queue()

        self.dispatch_thread = threading.Thread(target=self._dispatch_loop)
        # self.dispatch_thread.daemon = True  # else Ctrl-C will not work
        self.dispatch_thread.start()

    def _start_workers(self):
        procs = []
        for worker_index in range(self.num_workers):
            id_ = "worker_%d" % worker_index
            proc = run_worker(id_, self.workers_connect_to)
            procs.append((id_, proc))

        self.procs = procs

    def _setup_client_queue(self):
        self.queue_socket = self.context.socket(zmq.PUSH)
        self.queue_socket.connect(self.client_sends_to)

    def submit(self, data):
        self.queue_socket.send(b"PROCESS|%s" % data)

    def shutdown(self):
        self.queue_socket.send(b"DONE|")
        self.logger.info("wait for dispatch thread to join")
        #while self.dispatch_thread.isAlive():
        #    time.sleep(.001)

        # wait until all threads are finished
        self.dispatch_thread.join()
        self.queue_socket.close()

    def terminate_workers(self):
        for id_, p in self.procs:
            p.wait()
            self.logger.info("%s (pid=%d) finished" % (id_, p.pid))

    def _dispatch_loop(self):

        self.frontend = self.context.socket(zmq.PULL)
        self.frontend.bind(self.client_sends_to)

        self.backend = self.context.socket(zmq.ROUTER)
        self.backend.bind(self.workers_connect_to)

        poller = zmq.Poller()
        poller.register(self.backend, zmq.POLLIN)
        poller.register(self.frontend, zmq.POLLIN)

        available_workers = []
        self._wait_for_all_workers(available_workers, poller)
        self.logger.info("available workers: %d" % len(available_workers))

        self._dispatch_all_work(available_workers, poller)
        self._wait_until_all_workers_are_available(available_workers, poller)

        self._send_kill_to_all_workers(available_workers)
        self._terminate()

    def _wait_for_all_workers(self, available_workers, poller):
        missing = self.num_workers
        while missing:

            socks = dict(poller.poll())

            if socks.get(self.backend) == zmq.POLLIN:
                worker_id, empty, message = self.backend.recv_multipart()
                assert empty == b""
                assert message == b"READY"
                missing -= 1
                available_workers.append(worker_id)

        self.logger.info("all %d workers up" % len(available_workers))

    def _dispatch_all_work(self, available_workers, poller):

        while True:

            socks = dict(poller.poll())

            if socks.get(self.backend) == zmq.POLLIN:
                self._update_available_workers(available_workers)

            # we only check for data to process if we have at least one worker available:
            if socks.get(self.frontend) == zmq.POLLIN:
                if available_workers:
                    action, payload = self.frontend.recv().split(b"|", 1)
                    self.logger.info("dispatcher received: %s %r" % (action, payload))
                    if action == b"DONE":
                        break
                    else:
                        # send payload to be processed by one available worker:
                        worker_id = available_workers.pop()
                        msg = b"PROCESS|%s" % payload
                        self.backend.send_multipart([worker_id, b"", msg])
                else:
                    # got data but no worker available
                    time.sleep(.001)

    def _update_available_workers(self, available_workers):
        worker_id, empty, message = self.backend.recv_multipart()
        assert empty == b""
        assert message == b"READY"
        available_workers.append(worker_id)

    def _wait_until_all_workers_are_available(self, available_workers, poller):
        # we got KILL and wait until all workers are ready again:
        self.logger.info("num workers", self.num_workers)
        while len(available_workers) < self.num_workers:
            self.logger.info("avail workers", available_workers)
            socks = dict(poller.poll())
            if socks.get(self.backend) == zmq.POLLIN:
                self._update_available_workers(available_workers)
            else:
                time.sleep(.001)
        self.logger.info("all workers waiting")

    def _send_kill_to_all_workers(self, available_workers):

        while available_workers:
            worker_id = available_workers.pop()
            msg = b"KILL"
            self.backend.send_multipart([worker_id, b"", msg])
            self.logger.info("sent KILL to %s" % worker_id)

        self.logger.info("sent KILL to all workers")

    def _terminate(self):
        self.logger.info("terminate worker processes")
        self.terminate_workers()
        self.frontend.close()
        self.backend.close()
        self.logger.info("zmq frontend + backend closed")
