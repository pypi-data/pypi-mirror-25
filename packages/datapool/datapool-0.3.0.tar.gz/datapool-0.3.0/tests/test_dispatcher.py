# encoding: utf-8
from __future__ import print_function, division, absolute_import

from datapool.database import setup_db, dump_db
from datapool.dispatcher import Dispatcher, log_unexpected_exceptions
from datapool.utils import find_executable
from datapool.logger import logger


def test_log_unexpected_exceptions(regtest_logger):
    @log_unexpected_exceptions
    def by_zero(a, b, c):
        return a + b / c

    with regtest_logger():
        by_zero(1, 2, 0)
        by_zero("1", 2, c=0)
        by_zero({1}, b="2", c=0)


def _run_dispatcher(config, regtest_logger, lz, pathes):
    setup_db(config.db)
    with regtest_logger():
        dispatcher = Dispatcher(config, time_provider=lambda: "NOW")
        for path in pathes:
            for message in dispatcher.dispatch(path):
                logger().info("message: {}".format(message))


def test_dispatcher_ok(config, regtest, regtest_logger, lz):
    pathes = (
        "sites/example_site/site.yaml",
        "data/parameters.yaml",
        "data/FloDar/source_type.yaml",
        "data/FloDar/FloDar_python/source.yaml",
        "data/FloDar/FloDar_python/conversion.py",
        "data/FloDar/FloDar_python/raw_data/data-001.raw", )

    config.python.executable = find_executable("python")
    _run_dispatcher(config, regtest_logger, lz, pathes)
    dump_db(config.db, file=regtest, max_rows=25)


def test_dispatcher_no_python(config, regtest_logger, lz):
    # we don't configure r here to trigger an error
    pathes = ("data/FloDar/FloDar_python/conversion.py",
              "data/FloDar/FloDar_python/raw_data/data-001.raw")

    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_no_raw(config, regtest_logger, lz):
    lz.remove_file("data/FloDar/FloDar_python/raw_data/data-001.raw")
    pathes = ("data/FloDar/FloDar_python/conversion.py", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_no_script(config, regtest_logger, lz):
    lz.remove_file("data/FloDar/FloDar_python/conversion.py")
    pathes = ("data/FloDar/FloDar_python/raw_data/data-001.raw", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_multiple_scripts(config, regtest_logger, lz):
    lz.overwrite_file("data/FloDar/FloDar_python/conversion.jl",
                      lz.p("data/FloDar/FloDar_python/conversion.py"))
    pathes = ("data/FloDar/FloDar_python/raw_data/data-001.raw", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_wrong_yaml_name(config, regtest_logger, lz):
    pathes = ("data/invalid_name.yaml", )
    _run_dispatcher(config, regtest_logger, lz, pathes)


def test_dispatcher_invalid_yaml(config, regtest_logger, data_path, lz):
    lz.overwrite_file("data/parameters.yaml", data_path("yamls/empty.yaml"))
    pathes = ("data/parameters.yaml", )
    _run_dispatcher(config, regtest_logger, lz, pathes)
