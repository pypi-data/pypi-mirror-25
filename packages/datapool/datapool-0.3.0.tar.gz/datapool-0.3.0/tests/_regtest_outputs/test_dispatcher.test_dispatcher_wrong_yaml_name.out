INFO     eawag.datapool  connected to db sqlite+pysqlite:///<tmpdir>/datapool.db
INFO     eawag.datapool  message: don't know how to handle data/invalid_name.yaml
